File: comparison/java-gps/src/test/java/PointTest.java
import java.time.ZonedDateTime;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.hamcrest.Matchers.closeTo;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.jupiter.api.Assertions.*;


public class PointTest {

  private static final double TOLERANCE = 1e-5;
  private static final double HALF_CIRCUMFERENCE = Math.PI * Point.MEAN_EARTH_RADIUS;

  private ZonedDateTime time;
  private Point point;

  @BeforeEach
  public void perTestSetup() {
    time = ZonedDateTime.now();
    point = new Point(time, 1.5, 50.0, 25.0);
  }

  @Test
  @DisplayName("Time stored and retrieved correctly")
  public void time() {
    assertThat(point.getTime(), is(time));
  }

  @Test
  @DisplayName("Longitude stored and retrieved correctly")
  public void longitude() {
    assertThat(point.getLongitude(), closeTo(1.5, TOLERANCE));
  }

  @Test
  @DisplayName("Latitude stored and retrieved correctly")
  public void latitude() {
    assertThat(point.getLatitude(), closeTo(50.0, TOLERANCE));
  }

  @Test
  @DisplayName("Exception if longitude is too low")
  public void longitudeTooLow() {
    assertThrows(IllegalArgumentException.class, () -> new Point(time, -180.5, 0.0, 0.0));
  }

  @Test
  @DisplayName("Exception if longitude is too high")
  public void longitudeTooHigh() {
    assertThrows(IllegalArgumentException.class, () -> new Point(time, 180.5, 0.0, 0.0));
  }

  @Test
  @DisplayName("Exception if longitude is too low")
  public void latitudeTooLow() {
    assertThrows(IllegalArgumentException.class, () -> new Point(time, 0.0, -90.5, 0.0));
  }

  @Test
  @DisplayName("Exception if latitude is too high")
  public void latitudeTooHigh() {
    assertThrows(IllegalArgumentException.class, () -> new Point(time, 0.0, 90.5, 0.0));
  }

  @Test
  @DisplayName("Elevation stored and retrieved correctly")
  public void elevation() {
    assertThat(point.getElevation(), closeTo(25.0, TOLERANCE));
  }

  @Test
  @DisplayName("Correct string representation of a Point")
  public void pointAsString() {
    assertThat(point.toString(), is("(1.50000, 50.00000), 25.0 m"));
  }

  @Test
  @DisplayName("Correct distance between poles")
  public void distanceBetweenPoles() {
    Point northPole = new Point(time, 0.0, 90.0, 0.0);
    Point southPole = new Point(time, 0.0, -90.0, 0.0);
    assertAll(
      () -> assertThat(northPole.distanceTo(northPole), closeTo(0.0, TOLERANCE)),
      () -> assertThat(northPole.distanceTo(southPole), closeTo(HALF_CIRCUMFERENCE, TOLERANCE))
    );
  }

  @Test
  @DisplayName("Correct distance between opposing points on equator")
  public void distanceBetweenOpposingPointsOnEquator() {
    Point p1 = new Point(time, 0.0, 0.0, 0.0);
    Point p2 = new Point(time, 180.0, 0.0, 0.0);
    Point p3 = new Point(time, -180.0, 0.0, 0.0);
    assertAll(
      () -> assertThat(p1.distanceTo(p1), closeTo(0.0, TOLERANCE)),
      () -> assertThat(p1.distanceTo(p2), closeTo(HALF_CIRCUMFERENCE, TOLERANCE)),
      () -> assertThat(p1.distanceTo(p3), closeTo(HALF_CIRCUMFERENCE, TOLERANCE)),
      () -> assertThat(p2.distanceTo(p3), closeTo(0.0, TOLERANCE))
    );
  }
}

File: comparison/kotlin-gps/src/test/kotlin/PointTest.kt
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.assertions.withClue
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.doubles.plusOrMinus
import io.kotest.matchers.shouldBe
import java.time.ZonedDateTime
import kotlin.math.PI

@Suppress("unused")
class PointTest : StringSpec({
    val time = ZonedDateTime.now()
    val point = Point(time, 1.5, 50.0, 25.0)
    val tolerance = 1e-6
    val zero = 0.0 plusOrMinus tolerance
    val halfCircumference = (PI * Point.MEAN_EARTH_RADIUS) plusOrMinus tolerance

    "Properties have expected values" {
        withClue("Time") { point.time shouldBe time }
        withClue("Longitude") { point.longitude shouldBe (1.5 plusOrMinus tolerance) }
        withClue("Latitude") { point.latitude shouldBe (50.0 plusOrMinus tolerance) }
        withClue("Elevation") { point.elevation shouldBe (25.0 plusOrMinus tolerance) }
    }

    "Exception if longitude < -180 degrees" {
        shouldThrow<IllegalArgumentException> {
            Point(time, -180.5, 0.0, 0.0)
        }
    }

    "Exception if longitude > 180 degrees" {
        shouldThrow<IllegalArgumentException> {
            Point(time, 180.5, 0.0, 0.0)
        }
    }

    "Exception if latitude < -90 degrees" {
        shouldThrow<IllegalArgumentException> {
            Point(time, 0.0, -90.5, 0.0)
        }
    }

    "Exception if latitude > 90 degrees" {
        shouldThrow<IllegalArgumentException> {
            Point(time, 0.0, 90.5, 0.0)
        }
    }

    "Correct string representation" {
        point.toString() shouldBe "(1.50000, 50.00000), 25.0 m"
    }

    "Correct distance between poles" {
        val northPole = Point(time, 0.0, 90.0, 0.0)
        val southPole = Point(time, 0.0, -90.0, 0.0)
        withClue("North to itself") {
            northPole.distanceTo(northPole) shouldBe zero
        }
        withClue("North to South") {
            northPole.distanceTo(southPole) shouldBe halfCircumference
        }
    }

    "Correct distance hetween points on equator" {
        val p1 = Point(time, 0.0, 0.0, 0.0)
        val p2 = Point(time, 180.0, 0.0, 0.0)
        val p3 = Point(time, -180.0, 0.0, 0.0)

        withClue("To itself") { p1.distanceTo(p1) shouldBe zero }
        withClue("Opposing point (+180°)") { p1.distanceTo(p2) shouldBe halfCircumference }
        withClue("Opposing point (-180°)") { p1.distanceTo(p3) shouldBe halfCircumference }
        withClue("+180° to -180° (same point)") { p2.distanceTo(p3) shouldBe zero }
    }
})
